{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract 20k window around every significant SNP\n",
    "\n",
    "Routine to extract 20Kb windoes around every significant SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load database\n",
    "df = pd.read_csv(\"./significant_loci.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>P</th>\n",
       "      <th>zscore</th>\n",
       "      <th>trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>loc17357_pos75</td>\n",
       "      <td>9534</td>\n",
       "      <td>19598845</td>\n",
       "      <td>4.038489e-07</td>\n",
       "      <td>-614.2631</td>\n",
       "      <td>abaxial_trichomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>loc15949_pos141</td>\n",
       "      <td>9534</td>\n",
       "      <td>2850933</td>\n",
       "      <td>8.764997e-07</td>\n",
       "      <td>-639.4628</td>\n",
       "      <td>adaxial_trichomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>loc17815_pos41</td>\n",
       "      <td>9666</td>\n",
       "      <td>5105659</td>\n",
       "      <td>3.366722e-11</td>\n",
       "      <td>-403.2628</td>\n",
       "      <td>blade_area</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>loc17612_pos122</td>\n",
       "      <td>9666</td>\n",
       "      <td>639163</td>\n",
       "      <td>6.948563e-06</td>\n",
       "      <td>-641.4501</td>\n",
       "      <td>adaxial_trichomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>loc68546_pos257</td>\n",
       "      <td>59870</td>\n",
       "      <td>8709701</td>\n",
       "      <td>1.367609e-11</td>\n",
       "      <td>-402.3808</td>\n",
       "      <td>blade_area</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>loc68029_pos67</td>\n",
       "      <td>59870</td>\n",
       "      <td>2152335</td>\n",
       "      <td>4.121656e-06</td>\n",
       "      <td>-381.3673</td>\n",
       "      <td>teeth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>loc119342_pos200</td>\n",
       "      <td>63200</td>\n",
       "      <td>3462418</td>\n",
       "      <td>2.671646e-09</td>\n",
       "      <td>-609.3962</td>\n",
       "      <td>abaxial_trichomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>loc119365_pos112</td>\n",
       "      <td>63200</td>\n",
       "      <td>3809754</td>\n",
       "      <td>1.030289e-07</td>\n",
       "      <td>-612.9429</td>\n",
       "      <td>abaxial_trichomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>loc130774_pos89</td>\n",
       "      <td>63552</td>\n",
       "      <td>13207322</td>\n",
       "      <td>4.296808e-06</td>\n",
       "      <td>-381.4072</td>\n",
       "      <td>teeth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>loc130375_pos101</td>\n",
       "      <td>63552</td>\n",
       "      <td>4859480</td>\n",
       "      <td>6.457438e-06</td>\n",
       "      <td>-641.3799</td>\n",
       "      <td>adaxial_trichomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNP    CHR        BP             P    zscore  \\\n",
       "24    loc17357_pos75   9534  19598845  4.038489e-07 -614.2631   \n",
       "29   loc15949_pos141   9534   2850933  8.764997e-07 -639.4628   \n",
       "7     loc17815_pos41   9666   5105659  3.366722e-11 -403.2628   \n",
       "32   loc17612_pos122   9666    639163  6.948563e-06 -641.4501   \n",
       "4    loc68546_pos257  59870   8709701  1.367609e-11 -402.3808   \n",
       "14    loc68029_pos67  59870   2152335  4.121656e-06 -381.3673   \n",
       "18  loc119342_pos200  63200   3462418  2.671646e-09 -609.3962   \n",
       "22  loc119365_pos112  63200   3809754  1.030289e-07 -612.9429   \n",
       "15   loc130774_pos89  63552  13207322  4.296808e-06 -381.4072   \n",
       "31  loc130375_pos101  63552   4859480  6.457438e-06 -641.3799   \n",
       "\n",
       "                trait  \n",
       "24  abaxial_trichomes  \n",
       "29  adaxial_trichomes  \n",
       "7          blade_area  \n",
       "32  adaxial_trichomes  \n",
       "4          blade_area  \n",
       "14              teeth  \n",
       "18  abaxial_trichomes  \n",
       "22  abaxial_trichomes  \n",
       "15              teeth  \n",
       "31  adaxial_trichomes  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['CHR'].duplicated(keep=False)].sort_values(by=\"CHR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{59695: [80628],\n",
       " 61473: [1045000],\n",
       " 17365: [7371028],\n",
       " 61434: [8380992],\n",
       " 59870: [8709701, 2152335],\n",
       " 62359: [273498],\n",
       " 62189: [1927409],\n",
       " 9666: [5105659],\n",
       " 19888: [5505553],\n",
       " 47881: [2774374],\n",
       " 63432: [2670146],\n",
       " 61828: [5464180],\n",
       " 24083: [9244149],\n",
       " 63552: [13207322, 4859480],\n",
       " 39367: [232203],\n",
       " 15681: [2806641],\n",
       " 63200: [3462418, 3809754],\n",
       " 61248: [2151777],\n",
       " 62932: [4009105],\n",
       " 62050: [2398634],\n",
       " 9534: [19598845, 2850933],\n",
       " 23621: [2087672],\n",
       " 62125: [3731290],\n",
       " 60920: [7982493],\n",
       " 63323: [2650799],\n",
       " 4057: [1779621],\n",
       " 52789: [2593286],\n",
       " 61262: [4536745]}"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#checking which scaffolds contain more than one SNP\n",
    "scaffold_info = {}\n",
    "\n",
    "for idx, row in df.iterrows():\n",
    "    if row[\"CHR\"] in scaffold_info.keys(): \n",
    "        scaffold_info[row[\"CHR\"]].append(row[\"BP\"])\n",
    "    else:\n",
    "        scaffold_info[row[\"CHR\"]] = [row[\"BP\"]]\n",
    "        \n",
    "scaffold_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scaffold_4057_SNP_1779621\n",
      "AAAATATTCA 40001 1779621\n",
      "Scaffold_9534_SNP_19598845\n",
      "GATTACTAAG 40001 19598845\n",
      "Scaffold_9534_SNP_2850933\n",
      "ACTAGCTAAT 40001 2850933\n",
      "Scaffold_9666_SNP_5105659\n",
      "GACGACAATG 40001 5105659\n",
      "Scaffold_15681_SNP_2806641\n",
      "ATTCCATAAC 40001 2806641\n",
      "Scaffold_17365_SNP_7371028\n",
      "gtcctcgcct 40001 7371028\n",
      "Scaffold_19888_SNP_5505553\n",
      "ATTATTTCCA 40001 5505553\n",
      "Scaffold_23621_SNP_2087672\n",
      "TTCGAAAAGT 40001 2087672\n",
      "Scaffold_24083_SNP_9244149\n",
      "NNNNNNNNNN 40001 9244149\n",
      "Scaffold_39367_SNP_232203\n",
      "GGCATGATCC 40001 232203\n",
      "Scaffold_47881_SNP_2774374\n",
      "TAAAAATTTG 40001 2774374\n",
      "Scaffold_52789_SNP_2593286\n",
      "TCCTTGATCA 40001 2593286\n",
      "Scaffold_59695_SNP_80628\n",
      "TTTTATATAT 24180 80628\n",
      "Scaffold_59870_SNP_8709701\n",
      "TATTTTTGGA 40001 8709701\n",
      "Scaffold_59870_SNP_2152335\n",
      "ACCATGCAGG 40001 2152335\n",
      "Scaffold_60920_SNP_7982493\n",
      "GGTTCTTCTT 40001 7982493\n",
      "Scaffold_61248_SNP_2151777\n",
      "ACACCTCACC 40001 2151777\n",
      "Scaffold_61262_SNP_4536745\n",
      "TAATGCATGG 40001 4536745\n",
      "Scaffold_61434_SNP_8380992\n",
      "GTGTCATAGA 40001 8380992\n",
      "Scaffold_61473_SNP_1045000\n",
      "TCATCAAATA 40001 1045000\n",
      "Scaffold_61828_SNP_5464180\n",
      "TGGAGGATTT 40001 5464180\n",
      "Scaffold_62050_SNP_2398634\n",
      "NNNNNNNNNN 40001 2398634\n",
      "Scaffold_62125_SNP_3731290\n",
      "AATTGGGATA 40001 3731290\n",
      "Scaffold_62189_SNP_1927409\n",
      "TTTTTCTCTA 40001 1927409\n",
      "Scaffold_62359_SNP_273498\n",
      "CAATCTCGGA 40001 273498\n",
      "Scaffold_62932_SNP_4009105\n",
      "CCATGAGATC 40001 4009105\n",
      "Scaffold_63200_SNP_3462418\n",
      "TTACGGATTG 40001 3462418\n",
      "Scaffold_63200_SNP_3809754\n",
      "AAGAAAATTG 40001 3809754\n",
      "Scaffold_63323_SNP_2650799\n",
      "TTGTGCTTAT 40001 2650799\n",
      "Scaffold_63432_SNP_2670146\n",
      "TTTTAAAAAA 40001 2670146\n",
      "Scaffold_63552_SNP_13207322\n",
      "AAAGAGGATT 40001 13207322\n",
      "Scaffold_63552_SNP_4859480\n",
      "TCGATTTGAC 40001 4859480\n"
     ]
    }
   ],
   "source": [
    "# put in memory genome and perfome extraction of windows, \n",
    "# result should be in a multifasta, name will include chr_snp_trait\n",
    "\n",
    "#set genome file, for easy process, I am using the oneliner version of the genome\n",
    "genome_file = \"/home/carlos/GDRIVE/viburnumThings/Viburnum-Annotation/MAKER/reducedL50/genome/final_assembly_v3_oneline.fa\"\n",
    "window_size = 20000\n",
    "\n",
    "#produce output file\n",
    "with open(\"20kwindows.fasta\", \"a\") as file_window:\n",
    "\n",
    "    #open genome file\n",
    "    with open(genome_file, \"r\") as genome:\n",
    "\n",
    "        #read line by line\n",
    "        lines = genome.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            #check if current line is a header or a sequence\n",
    "            if line[0] == \">\":\n",
    "                scaffold_number = int(line[1:].strip().split(\";\")[0].split(\"_\")[1]) #get only scaffold number to match with CSV\n",
    "                continue #if is the header, do another loop to get the sequence\n",
    "            else:\n",
    "                if scaffold_number in scaffold_info.keys():\n",
    "                    for snp in scaffold_info[scaffold_number]:\n",
    "                        new_header = f\"Scaffold_{scaffold_number}_SNP_{snp}\" #set new header with more information\n",
    "\n",
    "                        seq = line.strip() #get the sequence\n",
    "\n",
    "                        #define limits of window and correct them if needed\n",
    "                        start_window = snp - window_size\n",
    "                        end_window = snp + window_size + 1\n",
    "                        if start_window < 0: start_window = 0\n",
    "                        if end_window > len(seq): end_window = len(seq)\n",
    "\n",
    "                        #get only the window around the snp\n",
    "                        window = seq[start_window:end_window]\n",
    "\n",
    "                        print(new_header)\n",
    "                        print(window[0:10], len(window), snp)\n",
    "\n",
    "                        #save every window as independent fasta in a multifasta file\n",
    "                        file_window.write(f\">{new_header}\\n{window}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extraction for trichome significant SNPs\n",
    "\n",
    "This is the same of 20kwindows but only for some snps significant for trichomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load database\n",
    "df = pd.read_csv(\"./trichome_branching_sig_scaffolds.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>loc104719_pos121</td>\n",
       "      <td>62540</td>\n",
       "      <td>3710393</td>\n",
       "      <td>margin_branch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>loc104619_pos172</td>\n",
       "      <td>62540</td>\n",
       "      <td>1173139</td>\n",
       "      <td>ab_branch+ad_branch+margin_branch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNP    CHR       BP                              trait\n",
       "18  loc104719_pos121  62540  3710393                      margin_branch\n",
       "19  loc104619_pos172  62540  1173139  ab_branch+ad_branch+margin_branch"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some scaffolds contain more than one snp, I need to consider this in the extraction\n",
    "df[df['CHR'].duplicated(keep=False)].sort_values(by=\"CHR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{3589: [9252545],\n",
       " 4506: [8488253],\n",
       " 5036: [12188081],\n",
       " 5277: [211152],\n",
       " 17057: [457058],\n",
       " 17618: [8223181],\n",
       " 19888: [3195966],\n",
       " 23621: [2202981],\n",
       " 28087: [43758310],\n",
       " 31062: [604045],\n",
       " 47900: [1597869],\n",
       " 51506: [1002237],\n",
       " 56543: [14798389],\n",
       " 61305: [2136144],\n",
       " 61434: [9194665],\n",
       " 61496: [980939],\n",
       " 62480: [2374442],\n",
       " 62491: [17612429],\n",
       " 62540: [3710393, 1173139],\n",
       " 62878: [7116797],\n",
       " 62942: [1452977],\n",
       " 63514: [229274]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaffold_info = {}\n",
    "\n",
    "for idx, row in df.iterrows():\n",
    "    if row[\"CHR\"] in scaffold_info.keys(): \n",
    "        scaffold_info[row[\"CHR\"]].append(row[\"BP\"])\n",
    "    else:\n",
    "        scaffold_info[row[\"CHR\"]] = [row[\"BP\"]]\n",
    "        \n",
    "scaffold_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scaffold_3589_SNP_9252545\n",
      "atacaatgac 40001 9252545\n",
      "Scaffold_4506_SNP_8488253\n",
      "CTTGCCGACA 40001 8488253\n",
      "Scaffold_5036_SNP_12188081\n",
      "TCAGCACCTG 40001 12188081\n",
      "Scaffold_5277_SNP_211152\n",
      "TCCTCGAAAT 40001 211152\n",
      "Scaffold_17057_SNP_457058\n",
      "AAACTATTGA 40001 457058\n",
      "Scaffold_17618_SNP_8223181\n",
      "tagttttgag 40001 8223181\n",
      "Scaffold_19888_SNP_3195966\n",
      "NNNNNNNNNN 40001 3195966\n",
      "Scaffold_23621_SNP_2202981\n",
      "GAGCTCGGGT 40001 2202981\n",
      "Scaffold_28087_SNP_43758310\n",
      "ATGTTATTTT 40001 43758310\n",
      "Scaffold_31062_SNP_604045\n",
      "CTCCCACAAA 40001 604045\n",
      "Scaffold_47900_SNP_1597869\n",
      "cgtactcagc 40001 1597869\n",
      "Scaffold_51506_SNP_1002237\n",
      "GATTCATGTG 40001 1002237\n",
      "Scaffold_56543_SNP_14798389\n",
      "GATTGTTGCA 40001 14798389\n",
      "Scaffold_61305_SNP_2136144\n",
      "AGCCCTCTAA 40001 2136144\n",
      "Scaffold_61434_SNP_9194665\n",
      "TTCCTATAAT 40001 9194665\n",
      "Scaffold_61496_SNP_980939\n",
      "ATTATGGGGG 40001 980939\n",
      "Scaffold_62480_SNP_2374442\n",
      "NNNNNNNNNN 40001 2374442\n",
      "Scaffold_62491_SNP_17612429\n",
      "AAAACAAAAG 40001 17612429\n",
      "Scaffold_62540_SNP_3710393\n",
      "ATAAGAGAAG 40001 3710393\n",
      "Scaffold_62540_SNP_1173139\n",
      "CCATATATTT 40001 1173139\n",
      "Scaffold_62878_SNP_7116797\n",
      "TCCTCCAGTT 40001 7116797\n",
      "Scaffold_62942_SNP_1452977\n",
      "ATAATTTTAA 40001 1452977\n",
      "Scaffold_63514_SNP_229274\n",
      "GCAAGGGATT 40001 229274\n"
     ]
    }
   ],
   "source": [
    "# put in memory genome and perfome extraction of windows, \n",
    "# result should be in a multifasta, name will include chr_snp_trait\n",
    "\n",
    "#set genome file, for easy process, I am using the oneliner version of the genome\n",
    "genome_file = \"/home/carlos/GDRIVE/viburnumThings/Viburnum-Annotation/MAKER/reducedL50/genome/final_assembly_v3_oneline.fa\"\n",
    "window_size = 20000\n",
    "\n",
    "#produce output file\n",
    "with open(\"20kwindows_trichomes.fasta\", \"a\") as file_window:\n",
    "\n",
    "    #open genome file\n",
    "    with open(genome_file, \"r\") as genome:\n",
    "\n",
    "        #read line by line\n",
    "        lines = genome.readlines()\n",
    "\n",
    "        for line in lines:\n",
    "            #check if current line is a header or a sequence\n",
    "            if line[0] == \">\":\n",
    "                scaffold_number = int(line[1:].strip().split(\";\")[0].split(\"_\")[1]) #get only scaffold number to match with CSV\n",
    "                continue #if is the header, do another loop to get the sequence\n",
    "            else:\n",
    "                if scaffold_number in scaffold_info.keys():\n",
    "                    for snp in scaffold_info[scaffold_number]:\n",
    "                        new_header = f\"Scaffold_{scaffold_number}_SNP_{snp}\" #set new header with more information\n",
    "\n",
    "                        seq = line.strip() #get the sequence\n",
    "\n",
    "                        #define limits of window and correct them if needed\n",
    "                        start_window = snp - window_size\n",
    "                        end_window = snp + window_size + 1\n",
    "                        if start_window < 0: start_window = 0\n",
    "                        if end_window > len(seq): end_window = len(seq)\n",
    "\n",
    "                        #get only the window around the snp\n",
    "                        window = seq[start_window:end_window]\n",
    "\n",
    "                        print(new_header)\n",
    "                        print(window[0:10], len(window), snp)\n",
    "\n",
    "                        #save every window as independent fasta in a multifasta file\n",
    "                        file_window.write(f\">{new_header}\\n{window}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
